### PR TITLE
feat: gateway graceful shutdown

### PR DESCRIPTION
**Pull request by scott-dn**
#125 

### DEMO:
image

### CODE REVIEW

**Comment by b0xtch** [comment]
Btw is this waiting for my PR #217 ?

**Comment by scott-dn** [comment]
> Btw is this blocked by my PR #217 ?

Not blocking, need a couple of polishes/refactors before ready to review.

File: crates/ursa-gateway/src/main.rs [hunk 1]
```diff
@@ -69,32 +79,112 @@ async fn main() -> Result<()> {
...
+                    };
+                    info!("Admin server stopped");
+                });
+                (worker, signal_rx)
+            };
+
+            let (ttl_cache_worker, mut ttl_cache_worker_signal_rx) = {
+                let mut shutdown_rx = shutdown_tx.subscribe();
+                let (signal_tx, signal_rx) = mpsc::channel(1);
+                let worker = spawn(async move {
```
Comment (b0xtch) [inline crates/ursa-gateway/src/main.rs:112]: Are these tasks aborted as a result of the signal? Abort will help the task finish then cancel the task iirc

Comment (b0xtch) [reply crates/ursa-gateway/src/main.rs:112]: I only the signal being sent out if an error to start. If I just send a sigint will they gracefully shutdown?

**Review by b0xtch — APPROVED** [review]
Prelim approval, sure the shutdown procedure could be improved

File: crates/ursa-gateway/src/main.rs (hunk 1 above)
Comment (scott-dn) [reply crates/ursa-gateway/src/main.rs:112]: > Are these tasks aborted as a result of the signal? Abort will help the task finish then cancel the task iirc

It's not only aborted but also waiting for tasks to be completed, in general, we don't want tasks to be ignored or lost.

Comment (scott-dn) [reply crates/ursa-gateway/src/main.rs:112]: > I only the signal being sent out if an error to start. If I just send a sigint will they gracefully shutdown?

It covers two scenarios:
1. if one critical task error starts, notify others tasks to stop immediately.
2. on SIGNIT, notify all tasks to finish all alive requests and then stop.

File: crates/ursa-gateway/src/main.rs [hunk 2]
```diff
@@ -69,32 +79,112 @@ async fn main() -> Result<()> {
...
+
+            select! {
+                _ = ctrl_c() => graceful_shutdown(shutdown_tx, sub_workers, main_shutdown_tx, main_worker).await,
+                _ = terminate => graceful_shutdown(shutdown_tx, sub_workers, main_shutdown_tx, main_worker).await,
+                _ = server_worker_signal_rx.recv() => graceful_shutdown(shutdown_tx, sub_workers, main_shutdown_tx, main_worker).await,
+                _ = admin_worker_signal_rx.recv() => graceful_shutdown(shutdown_tx, sub_workers, main_shutdown_tx, main_worker).await,
+                _ = ttl_cache_worker_signal_rx.recv() => graceful_shutdown(shutdown_tx, sub_workers, main_shutdown_tx, main_worker).await,
+                _ = worker_signal_rx.recv() => graceful_shutdown(shutdown_tx, sub_workers, main_shutdown_tx, main_worker).await
+            }
+            info!("Gateway shutdown successfully")
```
Comment (kckeiks) [inline crates/ursa-gateway/src/main.rs:166]:
```suggestion
            info!("Gateway shut down successfully")
```
shutdown is a noun

File: crates/ursa-gateway/src/main.rs [hunk 3]
```diff
@@ -69,32 +79,112 @@ async fn main() -> Result<()> {
...
+    sub_workers: Vec<JoinHandle<()>>,
+    main_shutdown_tx: mpsc::Sender<()>,
+    main_worker: JoinHandle<()>,
+) {
+    info!("Gateway shutting down...");
+    sub_shutdown_tx
+        .send(())
+        .expect("Send shutdown signal successfully");
+    for worker in sub_workers {
+        worker.await.expect("Worker to shutdown successfully");
```
Comment (kckeiks) [inline crates/ursa-gateway/src/main.rs:183]:
```suggestion
        worker.await.expect("Worker to shut down successfully");
```

File: crates/ursa-gateway/src/main.rs [hunk 4]
```diff
@@ -69,32 +79,112 @@ async fn main() -> Result<()> {
...
+    }
+    main_shutdown_tx
+        .send(())
+        .await
+        .expect("Send shutdown signal successfully");
+    main_worker.await.expect("Worker to shutdown successfully");
```
Comment (kckeiks) [inline crates/ursa-gateway/src/main.rs:189]:
```suggestion
    main_worker.await.expect("Worker to shut down successfully");
```

**Review by b0xtch — APPROVED** [review]

**Pull request merged**